vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLSocketHandlerImpl.java

```diff
@@ -0,0 +1,179 @@
+            subscribe(serverWebSocket, subscriptions, message);
+            break;
+          case STOP:
+            unsubscribe(serverWebSocket, subscriptions, opId);
+            break;
+          default:
+            sendError(serverWebSocket, opId, new Exception("Invalid message type!"));
+            break;
+        }
+      } catch (Exception e) {
+        sendError(serverWebSocket, null, e);
+      }
+    });
+
+    serverWebSocket.endHandler(v -> {
+      subscriptions.values().forEach(Subscription::cancel);
+
+      if (endHandler != null) {
+        endHandler.handle(serverWebSocket);
+      }
+    });
+  }
+
+  @Override
+  public GraphQLSocketHandler endHandler(Handler<ServerWebSocket> endHandler) {
+    this.endHandler = endHandler;
+
+    return this;
+  }
+
+  private void subscribe(
+    ServerWebSocket serverWebSocket,
+    Map<String, Subscription> subscriptions,
+    GraphQLMessageWithPayload message
+  ) {
+    String opId = message.getId();
+
+    // Unsubscribe if it's subscribed
+    if (subscriptions.containsKey(opId)) {
+      unsubscribe(serverWebSocket, subscriptions, opId);
+    }
+
+    GraphQLQuery payload = message.getPayload();
+    ExecutionInput.Builder builder = ExecutionInput.newExecutionInput();
+    builder.query(payload.getQuery());
+    String operationName = payload.getOperationName();
+    if (operationName != null) {
+      builder.operationName(operationName);
+    }
+    Map<String, Object> variables = payload.getVariables();
+    if (variables != null) {
+      builder.variables(variables);
+    }
+
+    graphQL.executeAsync(builder).thenAccept(executionResult -> {
+      Publisher<ExecutionResult> publisher = executionResult.getData();
+
+      AtomicReference<Subscription> subscriptionRef = new AtomicReference<>();
+      publisher.subscribe(new Subscriber<ExecutionResult>() {
+        @Override
+        public void onSubscribe(Subscription s) {
+          subscriptionRef.set(s);
+          subscriptions.put(opId, s);
+
+          s.request(1);
+        }
+
+        @Override
+        public void onNext(ExecutionResult er) {
+          sendMessage(serverWebSocket, opId, GraphQLMessage.Type.DATA, er);
+
+          subscriptionRef.get().request(1);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+          sendError(serverWebSocket, opId, t);
+        }
+
+        @Override
+        public void onComplete() {
+          sendMessage(serverWebSocket, opId, GraphQLMessage.Type.COMPLETE);
+          subscriptions.remove(opId);
+        }
+      });
+    });
+  }
+
+  private void unsubscribe(ServerWebSocket serverWebSocket, Map<String, Subscription> subscriptions, String opId) {
+    Subscription subscription = subscriptions.get(opId);
+
+    if (subscription != null) {
+      subscription.cancel();
+      subscriptions.remove(opId);
+    }
+  }
+
+  private void sendMessage(ServerWebSocket serverWebSocket, String opId, GraphQLMessage.Type type, ExecutionResult payload) {
```

Comment:
How about getting rid of the `GraphQLMessage` hierarchy and use `JsonObject`? It seems working with types does not simplify the code.